ledger/src/blockstore_db.rs

```diff
@@ -1130,14 +1130,21 @@ impl Database {
         batch.delete_range_cf::<C>(cf, from_index, to_index)
     }
 
+    /// Delete files whose slot range is within [from, to).
     pub fn delete_file_in_range_cf<C>(&self, from: Slot, to: Slot) -> Result<()>
     where
         C: Column + ColumnName,
     {
+        // Note that the default behavior of rocksdb's delete_file_in_range_cf
+        // deletes files within [from, to], while delete_range_cf applies to
+        // [from, to) instead.
+        //
+        // For consistency, we make our delete_file_in_range_cf works for
+        // [from, to) by adjusting the `to` slot range by 1.
         self.backend.delete_file_in_range_cf(
             self.cf_handle::<C>(),
             &C::key(C::as_index(from)),
-            &C::key(C::as_index(to)),
+            &C::key(C::as_index(to.saturating_sub(1))),
```

Comment:
After looking over the whole layers in solana-validator, i think we should make `delete_range_cf` take [from, to] instead.

note that `purge_special_columns_with_primary_index` isn't used anymore. I think you can tell that after reading the code. 

then the last caller of `delete_range_cf` would be `run_purge_with_stats`, so we can remove this awful abstraction leaking code there:

https://github.com/solana-labs/solana/blob/6d12bb6ec36dee3d46e0b530d7de6ca6d802a443/ledger/src/blockstore/blockstore_purge.rs#L164-L165

Reply:
I feel the same way after spotting `let to_slot = to_slot.saturating_add(1); `.  Let's do `[from, to]` for both!

Reply:
Turn out this is harder than I think as there're other purge functions that also need to be adjusted as well.

Reply:
@yhchiang-sol 

> Turn out this is harder than I think as there're other purge functions that also need to be adjusted as well.

yeah. i saw these modifications in this pr and it's fine by itself. however, as i said:

> note that purge_special_columns_with_primary_index isn't used anymore. I think you can tell that after reading the code.

actually, those modified code isn't used anymore. Sorry about that this terse comment... So, I created a real pr which is demonstrating what I'm trying to say: https://github.com/solana-labs/solana/pull/27529.